### PR TITLE
Image Customizer: Add documentation.

### PR DESCRIPTION
**Pull request by cwize1**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

Add some documentation on how to use the Mariner Image customizer. This includes a Getting Started guide, descriptions of the CLI arguments, and descriptions of the configuration file.

###### Change Log  

- Image Customizer: Add documentation.

###### Does this affect the toolchain?  

NO

###### Test Methodology

n/a

### CODE REVIEW

File: toolkit/tools/imagecustomizer/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+- Lower overhead, since you don't need to boot up and shutdown the OS.
+- More precision when making changes, since you won't see any side effects that come
+  from the OS running.
+- The image has fewer requirements (e.g. ssh doesn't need to be installed).
+
+Disadvantages:
+
+- Not all Linux tools place nicely when run under chroot. (Though most of the most
+  common tools do play nicely since they are used to build Linux images under chroot.)
+  So, some customizations aren't possible to do using MIC.
```
Comment (jiria) [inline toolkit/tools/imagecustomizer/README.md:27]: Consider adding some example here.

File: toolkit/tools/imagecustomizer/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+## Getting started
+
+1. Download a Mariner core image.
```
Comment (jiria) [inline toolkit/tools/imagecustomizer/README.md:31]: Consider adding a link and possibly expanding to "... or use any existing Mariner image."

File: toolkit/tools/imagecustomizer/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+   - `<config-file.yaml>`: The configuration file created in Step 2.
+
+   For a description of all the command line options, see:
+   [Mariner Image Customizer command line](./docs/cli.md)
+
+4. Use the customized image.
+
+   The customized image is placed in the file that you specified with the
+   `--output-image-file` parameter. You can now use this image as you see fit.
+   (For example, boot it in a Hyper-V VM.)
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/README.md:71]: Do you want to add a section for reporting issues?
Might be a good idea to create our own issue template in the future, but maybe for now adding a keyword in the issue title is needed

File: toolkit/tools/imagecustomizer/docs/cli.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+The base image file to customize.
+
+Supported image file formats: vhd, vhdx, qcow2, and raw.
+
+## --output-image-file=FILE-PATH
+
+Required.
+
+The image format of the output customized image file.
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/cli.md:25]: I expected the output format to be appended when I tried out the tool, but it was not. so, you may want to add an example here or mention that this should include the image extension.

File: toolkit/tools/imagecustomizer/docs/cli.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+given the lowest priority.
+
+## --disable-base-image-rpm-repos
+
+Disable the base image's installed RPM repos as a source of RPMs during package
+installation.
+
+## --log-level=LEVEL
+
+The verbosity of logs the tool outputs.
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/cli.md:78]: mention the default?

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+   4. Update packages ([PackageListsUpdate](#packagelistsupdate-string),
+   [PackagesUpdate](#packagesupdate-string))
+
+3. Update hostname. ([Hostname](#hostname-string))
+
+4. Copy additional files. ([AdditionalFiles](#additionalfiles-mapstring-fileconfig))
+
+5. Add/update users. ([Users](#users-user))
+
+6. Enable/disable services. ([Services](#services))
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/configuration.md:28]: kernel modules missing from the list.

File: toolkit/tools/imagecustomizer/docs/cli.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+The image format of the output customized image file.
+
+Options: vhd, vhdx, qcow2, and raw.
+
+## --output-image-format=FORMAT
+
+Required.
+
+he image format of the output customized image file.
```
Comment (jiria) [inline toolkit/tools/imagecustomizer/docs/cli.md:33]: nit: T is missing

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+
+Top-level type: [Config](#config)
+
+Operation ordering:
+
+1. Override the `/etc/resolv.conf` file with the version from the host OS.
+
+2. Update packages:
+
+   1. Remove packages ([PackageListsRemove](#packagelistsremove-string),
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/configuration.md:11]: Do we want to add an explicit note on changing installed kernel?

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+
+## FileConfig
+
+Specifies options for placing a file in the OS.
+
+Type is used by: [AdditionalFiles](#additionalfiles-mapstring-fileconfig)
+
+### Path [string]
+
+The absolute path of the destination file.
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/configuration.md:64]: maybe add a note that same file can have multiple destinations?

File: toolkit/tools/imagecustomizer/docs/cli.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+## --output-image-file=FILE-PATH
+
+Required.
+
+The image format of the output customized image file.
+
+Options: vhd, vhdx, qcow2, and raw.
+
+## --output-image-format=FORMAT
```
Comment (jiria) [inline toolkit/tools/imagecustomizer/docs/cli.md:29]: Left this comment elsewhere, sorry have not yet caught up. But, do we really need the output-image-format when similar information is passed through output-image-file?

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+Options for configuring systemd services.
+
+```yaml
+SystemConfig:
+  Services:
+    Enable:
+    - sshd
+```
+
+## User
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/configuration.md:379]: duplicate to above ###Users

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 9 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md:64]: That is a feature of the `AdditionalFiles` property, not the `FileConfig` object,

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 11 above)
Comment (romoh) [reply toolkit/tools/imagecustomizer/docs/configuration.md:379]: looks like "Services" is the one out of context here

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+
+```yaml
+SystemConfig:
+  Users:
+  - Name: test
+    SecondaryGroups:
+    - sudo
+```
+
+### StartupCommand [string]
```
Comment (romoh) [inline toolkit/tools/imagecustomizer/docs/configuration.md:528]: Is this implemented yet?

File: toolkit/tools/imagecustomizer/README.md (hunk 3 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/README.md:71]: Probably not in the public documentation.

**Review by romoh — CHANGES_REQUESTED** [review]
some comments

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+Same as [PackagesUpdate](#packagesupdate-string) but the packages are specified in a
+separate YAML (or JSON) file.
+
+The other YAML file schema is specified by [PackageList](#packagelist).
+
+Example:
+
+```yaml
+SystemConfig:
+  PackageListsRemove:
```
Comment (jiria) [inline toolkit/tools/imagecustomizer/docs/configuration.md:278]: PackageListsUpdate?

File: toolkit/tools/imagecustomizer/docs/configuration.md [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+
+4. Copy additional files. ([AdditionalFiles](#additionalfiles-mapstring-fileconfig))
+
+5. Add/update users. ([Users](#users-user))
+
+6. Enable/disable services. ([Services](#services))
+
+7. Run post-install scripts. ([PostInstallScripts](#postinstallscripts-script))
+
+8. Run post-finalize scripts. ([FinalizeImageScripts](#finalizeimagescripts-script))
```
Comment (jiria) [inline toolkit/tools/imagecustomizer/docs/configuration.md:32]: Run finalize scripts?

File: toolkit/tools/imagecustomizer/docs/cli.md (hunk 7 above)
Comment (elainezhao96) [inline toolkit/tools/imagecustomizer/docs/cli.md:33]: nit: The

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 6 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md:28]: Thanks

File: toolkit/tools/imagecustomizer/README.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+   For documentation on the supported configuration options, see:
+   [Mariner Image Customizer configuration](./docs/configuration.md)
+
+3. Run the Mariner Image Customizer tool.
+
+   For example:
+
+    ```bash
+    sudo ./imagecustomizer \
```
Comment (elainezhao96) [inline toolkit/tools/imagecustomizer/README.md:51]: maybe I missed some context, is imagecustomizer an output executable of `go build`? is this already in this dir or user download it from somewhere?

File: toolkit/tools/imagecustomizer/README.md [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+Advantages:
+
+- Lower overhead, since you don't need to boot up and shutdown the OS.
+- More precision when making changes, since you won't see any side effects that come
+  from the OS running.
+- The image has fewer requirements (e.g. ssh doesn't need to be installed).
+
+Disadvantages:
+
+- Not all Linux tools place nicely when run under chroot. (Though most of the most
```
Comment (fintelia) [inline toolkit/tools/imagecustomizer/README.md:25]: nit: place -> play

File: toolkit/tools/imagecustomizer/README.md (hunk 2 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/README.md:31]: There aren't any public links to the Mariner image file (e.g. vhd, vhdx) downloads.

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 11 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md:430]: `User` is a type. `Users` is a property on the `SystemConfig` type.

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 12 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md:579]: Yes. The implementation was taken from the imager tool.

File: toolkit/tools/imagecustomizer/docs/cli.md (hunk 7 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/cli.md]: Fixed.

File: toolkit/tools/imagecustomizer/docs/cli.md (hunk 5 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/cli.md:78]: Fixed.

File: toolkit/tools/imagecustomizer/README.md (hunk 15 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/README.md:52]: Yes. You can build this executable by running `make -C toolkit go-imagecustomizer`.

The intention is to include these docs in the same directory as the `imagecustomizer` binary.

File: toolkit/tools/imagecustomizer/docs/cli.md (hunk 10 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/cli.md:27]: I think the problem is that some of the image formats don't have a standard filename extension. Raw images are the prime example of this.

File: toolkit/tools/imagecustomizer/README.md (hunk 16 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/README.md]: Fixed.

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 13 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md]: Fixed.

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 8 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md:12]: I added a section about replacing packages. And included the kernel package as an example.

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 11 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md]: I added `type` to the titles of the types to try to make this clearer.

File: toolkit/tools/imagecustomizer/docs/configuration.md (hunk 14 above)
Comment (cwize1) [reply toolkit/tools/imagecustomizer/docs/configuration.md]: Fixed.

File: toolkit/tools/imagecustomizer/README.md (hunk 2 above)
Comment (jiria) [reply toolkit/tools/imagecustomizer/README.md:32]: Ok, should we still mention that any already customized image can be used as an input as well?

**Review by jiria — APPROVED** [review]

**Review by romoh — APPROVED** [review]

**Pull request merged**